### PR TITLE
Avoid SHGetSpecialFolderPathA in UWP.

### PR DESCRIPTION
**Pull request by jaykrell**
SHGetSpecialFolderPathA was introduced:

 commit 66c02c7826ed7b65ddbd071c6ac0a86b1407ebc4
 Author: Nick Mathewson <nickm@torproject.org>
 Date:   Fri Jan 8 04:02:19 2010 -0500

    Look at the proper /etc/hosts file on windows.

There are many nuances and possibilities here.

 - A UWP port is just starting, and might not succeed, so this
   is speculative.

 - Reading the hosts file appears optional, so we could just skip it.

 - There are many ways to get the system directory, as the original PR
   indicated, and this might not be honoring the original intent.
   i.e. Maybe should use GetWindowsDirectory and then append system32.
   Or use environment variables %Windir% or %SystemRoot%.

 - The #if surely is not needed, but one UWP-compatible way
   could be used for desktop apps.

 - #if UWP is my own invention.
   shlobj.h itself uses #if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)
   which feels unwieldy.

 - If this is the only use of shell32.dll, then maybe don't use shell32.lib.

I don't know why neither GetWindowsDiretory nor GetSystemDiretory were used in the first place.
I'd rather use kernel32.dll than shell32.dll.

### CODE REVIEW

File: evdns.c [hunk 1]
```diff
@@ -3643,6 +3645,13 @@ evdns_get_default_hosts_filename(void)
 	char path[MAX_PATH+1];
 	static const char hostfile[] = "\\drivers\\etc\\hosts";
 	char *path_out;
+#if UWP
```
Comment (azat) [inline evdns.c:3648]: Is it defined somewhere? Anyway it is too short, verbose is good in this case, let's use the `WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP)`

File: evdns.c [hunk 2]
```diff
@@ -3643,6 +3645,13 @@ evdns_get_default_hosts_filename(void)
 	char path[MAX_PATH+1];
 	static const char hostfile[] = "\\drivers\\etc\\hosts";
 	char *path_out;
+#if UWP
+	/* Or GetWindowsDirectoryA + system32? */
+	size_t len_out = GetSystemDirectoryA(path, MAX_PATH);
+	if (len_out == 0 || len_out >= MAX_PATH)
+		return NULL;
+	len_out += sizeof(hostfile);
```
Comment (azat) [inline evdns.c:3653]: You don't append `hostfile` to it, and no return statement

File: evdns.c (hunk 1 above)
Comment (jaykrell) [reply evdns.c:3648]: SHGetSpecialFolderPathA is not available in UWP, per its documentation.

I placed the endif too late, oops, right.

kernel32.dll is not loaded dynamically here as you describe.

You want both GetSystemDirectory and GetWindowDirectory + system32?

Comment (azat) [reply evdns.c:3648]: >You want both GetSystemDirectory and GetWindowDirectory + system32?

Yes, but only if the result may differ

**Comment by azat** [comment]
@ygj6 can you take a look? If no - just unassign your self

File: evdns.c (hunk 2 above)
Comment (azat) [reply evdns.c:3653]: @jaykrell do you have time to finish this?

File: evdns.c (hunk 1 above)
Comment (azat) [reply evdns.c:3648]: >I placed the endif too late, oops, right.

This one does not fixed either

**Pull request closed without merging**